File: songs/ruby/22-unique-d.rb
require_relative 'MySongs'
include Songs

def unique(tags)
  if tags.length() <= 1 then
    tags
  else
    head, *tail = tags
    tail.delete(head)
    return [head] + unique(tail)
  end
end

tags = SONGS
  .map { |song| song['tags'] }
  .compact.flatten

puts "#{unique(tags)}"

File: songs/php/22-unique-d.php
<?php
require_once(__DIR__.'/MyClassSongs.php');

function unique($tags) {
  if(sizeof($tags) <= 1) return $tags;
  else {
    $head = array_shift($tags);
    $exclude = array_values(
      array_diff($tags, [$head]));
    return array_merge(
      [$head], unique($exclude)
    );
  }
}

$tagss = array_map(
  function ($song) { return $song->tags; },
  $songs);

$tags = array_merge(...$tagss);
$tags = unique($tags);

printf("%s\n", json_encode($tags));
